src/common/webrequest_curl.cpp

```diff
@@ -0,0 +1,559 @@
+        static_cast<const char*>(authStr.mb_str()));
+    m_request.StartRequest();
+}
+
+//
+// wxWebSessionCURL
+//
+
+int wxWebSessionCURL::ms_activeSessions = 0;
+
+wxWebSessionCURL::wxWebSessionCURL() :
+    m_handle(NULL),
+    m_condition(m_mutex),
+    m_shuttingDown(false)
+{
+    // Initialize CURL globally if no sessions are active
+    if ( ms_activeSessions == 0 )
+        InitializeCURL();
+
+    ms_activeSessions++;
+}
+
+wxWebSessionCURL::~wxWebSessionCURL()
+{
+    {
+        // Notify the work thread
+        m_shuttingDown = true;
+        wxMutexLocker lock(m_mutex);
+        m_condition.Broadcast();
+    }
+
+    // Wait for work thread to finish
+    if ( GetThread() && GetThread()->IsRunning() )
+        GetThread()->Wait(wxTHREAD_WAIT_BLOCK);
+
+    if ( m_handle )
+        curl_multi_cleanup(m_handle);
+
+    // Global CURL cleanup if this is the last session
+    --ms_activeSessions;
+    if ( ms_activeSessions == 0 )
+        CleanupCURL();
+}
+
+void wxWebSessionCURL::Initialize()
+{
+    m_handle = curl_multi_init();
+}
+
+wxWebRequest* wxWebSessionCURL::CreateRequest(const wxString& url, int id)
+{
+    if ( !m_handle )
+        Initialize();
+
+    return new wxWebRequestCURL(*this, id, url);
+}
+
+static CURLMcode wx_curl_multi_wait(CURLM *multi_handle, int timeout_ms,
+    int *ret)
+{
+    // since libcurl 7.28.0 the curl_multi_wait method is more convient than
+    // calling multiple curl_multi_... methods.
+    // When support for older libcurl versions is dropped this wrapper can be
+    // removed.
+#if wxCURL_HAVE_MULTI_WAIT
+    return curl_multi_wait(multi_handle, NULL, 0, timeout_ms, ret);
+#else
+    wxASSERT(ret != NULL);
+
+    fd_set fdread;
+    fd_set fdwrite;
+    fd_set fdexcep;
+    timeval timeout;
+
+    long curl_timeo;
+
+    curl_multi_timeout(multi_handle, &curl_timeo);
+    if ( curl_timeo < 0 )
+        curl_timeo = timeout_ms;
+
+    timeout.tv_sec = curl_timeo / 1000;
+    timeout.tv_usec = (curl_timeo % 1000) * 1000;
+
+    FD_ZERO(&fdread);
+    FD_ZERO(&fdwrite);
+    FD_ZERO(&fdexcep);
+
+    curl_multi_fdset(multi_handle, &fdread, &fdwrite, &fdexcep, ret);
+    if ( *ret == -1 )
+        return CURLM_OK;
+    else if ( select(*ret + 1, &fdread, &fdwrite, &fdexcep, &timeout) == -1 )
+        return CURLM_BAD_SOCKET;
+    else
+        return CURLM_OK;
+#endif
+}
+
+wxThread::ExitCode wxWebSessionCURL::Entry()
+{
+    m_mutex.Lock();
+
+    int activeRequests = -1;
+    int repeats = 0;
+
+    while ( activeRequests )
+    {
+        // Handle cancelled requests
+        {
+            wxMutexLocker lock(m_cancelledMutex);
+            while ( !m_cancelledRequests.empty() )
+            {
+                wxObjectDataPtr<wxWebRequestCURL> request(m_cancelledRequests.back());
+                m_cancelledRequests.pop_back();
+                curl_multi_remove_handle(m_handle, request->GetHandle());
+                request->SetState(wxWebRequest::State_Cancelled);
```

Comment:
This doesn't seem to be MT-safe. There should be some mutex/critical section protecting the state explicitly, otherwise I don't see what prevents the main thread from modifying it "at the same time".

Reply:
This, and several other MT-safety problems, were fixed now.